python/pyspark/pandas/internal.py

```diff
@@ -1143,7 +1143,10 @@ def restore_index(
             drop = index_field not in data_columns
             pdf = pdf.set_index(index_field, drop=drop, append=append)
             append = True
-        pdf = pdf[data_columns]
+
+        # SPARK-37930: Fix DataFrame select subset with duplicated columns
+        # Remove duplicated columns before select `data_columns`
+        pdf = pdf.loc[:, ~pdf.columns.duplicated()][data_columns]
```

Comment:
ohh. sorry I missed https://spark.apache.org/docs/latest/api/python/user_guide/pandas_on_spark/best_practices.html?highlight=duplicate#do-not-use-duplicated-column-names. Actually pandas API on Spark does not officially support duplicated column names for now.

But It think we can fix this as the fix seems minimized. BTW, can we do `pdf[~pdf.columns.duplicated()][data_columns]`?

Reply:
Otherwise LGTM

Reply:
> can we do `pdf[~pdf.columns.duplicated()][data_columns]`?

I think `pdf[~pdf.columns.duplicated()]` is used to select rows. Do you mind if I keep using `.loc`

Reply:
Should we change the [docs](https://spark.apache.org/docs/latest/api/python/user_guide/pandas_on_spark/best_practices.html?highlight=duplicate#do-not-use-duplicated-column-names)  in this PR

Reply:
>  Actually pandas API on Spark does not officially support duplicated column names for now.

learned

> Should we change the docs in this PR
```
        import pyspark.pandas as ps
        psdf = ps.DataFrame({'a': [1, 2], 'b': [3, 4]})
        psdf.columns = ["a", "a"]
        print(psdf)
```
FYI, this still raise the exception `pyspark.sql.utils.AnalysisException: Reference 'a' is ambiguous, could be: a, a.`.

And sorry late to say, I think this PR is only for `self combine` case, right? so another option for original PR, since we have made it clear in the document that the same column is not supported, should we directly throw an exception when self combine?

Reply:
@Yikun Thanks for the clarification. I prefer to change behavior to follow pandas, but do not mind if we decide to keep this difference.

Reply:
Maybe throwing an exception is an idea too .. actually we have a lot of places like this, not only here. so let's just fix this it first, and revisit the duplicate names later separately.